### DISCUSSION TITLE
How to return list of transitions and states that are allowed from the current state

### DISCUSSION THREAD

**Discussion by nivekph** [statelyai/xstate#1411] [General]
```js
const timeOfDayMachine = Machine({
  id: 'timeOfDay',
  initial: 'unknown',
  context: {
    time: undefined
  },
  states: {
    // Transient state
    unknown: {
      on: {
        'ACTION': [
          { target: 'morning', cond: 'isBeforeNoon' },
          { target: 'afternoon', cond: 'isBeforeSix' },
          { target: 'evening' }
        ], 
        'ACTION2': "morning"
      }
    },
    morning: {},
    afternoon: {},
    evening: {}
  }
}, {
  guards: {
    isBeforeNoon: ()=> {},
    isBeforeSix: () => {}
  }
});
// currentState is unknown
// return list of transitions that are allowed from the current state 
// => ["ACTION", "ACTION2"]

// return list of states that are allowed from the current state 
// => 'isBeforeNoon'  -> ["morning"]
// => 'isBeforeSix'  ->  ["afternoon", "morning"]
// =>  ->  ["evening", "morning"]
```

**Comment by MrNovado** [comment]
@penghap 
As far as I know, `xstate` model (`Schema | Context | Event`) does not allow you to describe those restrictions currently. Also, no idea whether or not those are planed, but it would be neat to be able to do that out of the box.

I made a proposal some time ago about related topic:
https://github.com/davidkpiano/xstate/discussions/1056

And a proof of concept tool to generate those `state2EventMap` models:
https://github.com/davidkpiano/xstate/discussions/1056#discussioncomment-29846
https://github.com/MrNovado/xstate-gentype-proof/blob/v1.0.0-simple/src/simple.machine.gen.ts#L160

And you can make those maps yourself, but then again they won't be coupled with your `xstate-Schema` and their maintenance will be entirely on you.

**Reply by mattpocock** [reply]
@MrNovado @penghap This falls under the remit of what [xstate-codegen](https://github.com/mattpocock/xstate-codegen) is trying to accomplish. If you're interested, the beta is available on npm under the 'next' flag.
